src/python/pants/rules/core/fmt.py

```diff
@@ -22,6 +23,16 @@ class FmtResult:
 class TargetWithSources:
   """A union for registration of a formattable target type."""
 
+  @staticmethod
+  def valid_target(target_adaptor: TargetAdaptor, *, union_membership: UnionMembership) -> bool:
```

Comment:
Good point! I'm taking your suggestion from the other comment to call it `is_formattable_and_lintable`. A bit verbose, but very explicit. `has_sources` is a misnomer because this also checks for union membership.

Reply:
Works for me, but FYI the reason I was happy with `has_sources` is that this method would actually be called `pants.rules.core.fmt.TargetWithSources.has_sources`

`TargetWithSources.has_sources` or `fmt.TargetWithSources.has_sources` I think encompass the salient info reasonably, but if in doubt, verbosity isn't a bad thing :)